File: server/dto/alert.go
package dto

import (
	"time"

	"github.com/mingrenya/AI-Waf/server/model"
)

// AlertChannel DTO

// CreateAlertChannelRequest 创建告警渠道请求
type CreateAlertChannelRequest struct {
	Name    string                 `json:"name" binding:"required"`
	Type    string                 `json:"type" binding:"required,oneof=webhook slack discord dingtalk wecom"`
	Config  map[string]interface{} `json:"config" binding:"required"`
	Enabled bool                   `json:"enabled"`
}

// UpdateAlertChannelRequest 更新告警渠道请求
type UpdateAlertChannelRequest struct {
	Name    string                 `json:"name"`
	Config  map[string]interface{} `json:"config"`
	Enabled *bool                  `json:"enabled"`
}

// AlertChannelResponse 告警渠道响应
type AlertChannelResponse struct {
	ID        string                 `json:"id"`
	Name      string                 `json:"name"`
	Type      string                 `json:"type"`
	Config    map[string]interface{} `json:"config"`
	Enabled   bool                   `json:"enabled"`
	CreatedAt time.Time              `json:"createdAt"`
	UpdatedAt time.Time              `json:"updatedAt"`
}

// TestAlertChannelRequest 测试告警渠道请求
type TestAlertChannelRequest struct {
	Message string `json:"message" binding:"required"`
}

// AlertRule DTO

// CreateAlertRuleRequest 创建告警规则请求
type CreateAlertRuleRequest struct {
	Name        string                   `json:"name" binding:"required"`
	Description string                   `json:"description"`
	Conditions  []model.AlertCondition   `json:"conditions" binding:"required,min=1,dive"`
	Logic       string                   `json:"logic" binding:"required,oneof=AND OR"`
	Channels    []string                 `json:"channels" binding:"required,min=1"`
	Template    string                   `json:"template" binding:"required"`
	Cooldown    int                      `json:"cooldown" binding:"required,min=1"`
	Severity    string                   `json:"severity" binding:"required,oneof=low medium high critical"`
	Enabled     bool                     `json:"enabled"`
}

// UpdateAlertRuleRequest 更新告警规则请求
type UpdateAlertRuleRequest struct {
	Name        string                   `json:"name"`
	Description string                   `json:"description"`
	Conditions  []model.AlertCondition   `json:"conditions"`
	Logic       string                   `json:"logic" binding:"omitempty,oneof=AND OR"`
	Channels    []string                 `json:"channels"`
	Template    string                   `json:"template"`
	Cooldown    *int                     `json:"cooldown"`
	Severity    string                   `json:"severity" binding:"omitempty,oneof=low medium high critical"`
	Enabled     *bool                    `json:"enabled"`
}

// AlertRuleResponse 告警规则响应
type AlertRuleResponse struct {
	ID          string                   `json:"id"`
	Name        string                   `json:"name"`
	Description string                   `json:"description"`
	Conditions  []model.AlertCondition   `json:"conditions"`
	Logic       string                   `json:"logic"`
	Channels    []string                 `json:"channels"`
	Template    string                   `json:"template"`
	Cooldown    int                      `json:"cooldown"`
	Severity    string                   `json:"severity"`
	Enabled     bool                     `json:"enabled"`
	CreatedAt   time.Time                `json:"createdAt"`
	UpdatedAt   time.Time                `json:"updatedAt"`
	CreatedBy   string                   `json:"createdBy,omitempty"`
}

// AlertHistory DTO

// AlertHistoryResponse 告警历史响应
type AlertHistoryResponse struct {
	ID              string                 `json:"id"`
	RuleID          string                 `json:"ruleId"`
	RuleName        string                 `json:"ruleName"`
	Severity        string                 `json:"severity"`
	Message         string                 `json:"message"`
	Details         map[string]interface{} `json:"details"`
	Channels        []string               `json:"channels"`
	Status          string                 `json:"status"`
	ErrorMessage    string                 `json:"errorMessage,omitempty"`
	TriggeredAt     time.Time              `json:"triggeredAt"`
	SentAt          *time.Time             `json:"sentAt,omitempty"`
	AcknowledgedAt  *time.Time             `json:"acknowledgedAt,omitempty"`
	AcknowledgedBy  string                 `json:"acknowledgedBy,omitempty"`
}

// GetAlertHistoryRequest 获取告警历史请求
type GetAlertHistoryRequest struct {
	RuleID    string    `form:"ruleId"`
	Severity  string    `form:"severity" binding:"omitempty,oneof=low medium high critical"`
	Status    string    `form:"status" binding:"omitempty,oneof=pending sent failed acknowledged"`
	StartTime time.Time `form:"startTime"`
	EndTime   time.Time `form:"endTime"`
	Page      int       `form:"page" binding:"min=1"`
	PageSize  int       `form:"pageSize" binding:"min=1,max=100"`
}

// AcknowledgeAlertRequest 确认告警请求
type AcknowledgeAlertRequest struct {
	Comment string `json:"comment"`
}

// AlertTemplate DTO

// CreateAlertTemplateRequest 创建告警模板请求
type CreateAlertTemplateRequest struct {
	Name        string   `json:"name" binding:"required"`
	Description string   `json:"description"`
	Content     string   `json:"content" binding:"required"`
	Variables   []string `json:"variables"`
}

// UpdateAlertTemplateRequest 更新告警模板请求
type UpdateAlertTemplateRequest struct {
	Name        string   `json:"name"`
	Description string   `json:"description"`
	Content     string   `json:"content"`
	Variables   []string `json:"variables"`
}

// AlertTemplateResponse 告警模板响应
type AlertTemplateResponse struct {
	ID          string    `json:"id"`
	Name        string    `json:"name"`
	Description string    `json:"description"`
	Content     string    `json:"content"`
	Variables   []string  `json:"variables"`
	IsBuiltIn   bool      `json:"isBuiltIn"`
	CreatedAt   time.Time `json:"createdAt"`
	UpdatedAt   time.Time `json:"updatedAt"`
	CreatedBy   string    `json:"createdBy,omitempty"`
}

// Alert Statistics DTO

// AlertStatisticsResponse 告警统计响应
type AlertStatisticsResponse struct {
	TotalAlerts       int64                     `json:"totalAlerts"`
	AlertsBySeverity  map[string]int64          `json:"alertsBySeverity"`
	AlertsByStatus    map[string]int64          `json:"alertsByStatus"`
	TopAlertRules     []TopAlertRule            `json:"topAlertRules"`
	RecentAlerts      []AlertHistoryResponse    `json:"recentAlerts"`
}

// TopAlertRule Top 告警规则
type TopAlertRule struct {
	RuleID    string `json:"ruleId"`
	RuleName  string `json:"ruleName"`
	Count     int64  `json:"count"`
	Severity  string `json:"severity"`
}

// ChannelConfig 渠道配置接口 - 用于不同渠道的配置验证

// WebhookConfig Webhook 配置
type WebhookConfig struct {
	URL     string            `json:"url" binding:"required,url"`
	Method  string            `json:"method" binding:"required,oneof=GET POST PUT"`
	Headers map[string]string `json:"headers"`
	Timeout int               `json:"timeout" binding:"min=1,max=300"` // 超时时间（秒）
}

// SlackConfig Slack 配置
type SlackConfig struct {
	WebhookURL string `json:"webhookUrl" binding:"required,url"`
	Channel    string `json:"channel"`
	Username   string `json:"username"`
	IconEmoji  string `json:"iconEmoji"`
}

// DiscordConfig Discord 配置
type DiscordConfig struct {
	WebhookURL string `json:"webhookUrl" binding:"required,url"`
	Username   string `json:"username"`
	AvatarURL  string `json:"avatarUrl"`
}

// DingTalkConfig 钉钉配置
type DingTalkConfig struct {
	WebhookURL string `json:"webhookUrl" binding:"required,url"`
	Secret     string `json:"secret"`
	AtMobiles  []string `json:"atMobiles"`
	IsAtAll    bool   `json:"isAtAll"`
}

// WeComConfig 企业微信配置
type WeComConfig struct {
	WebhookURL string `json:"webhookUrl" binding:"required,url"`
	MentionedList []string `json:"mentionedList"`
	MentionedMobileList []string `json:"mentionedMobileList"`
}

File: web/src/types/alert.ts
// 告警相关类型定义

/**
 * 告警通道类型
 */
export enum AlertChannelType {
    Webhook = "webhook",
    Slack = "slack",
    Discord = "discord",
    DingTalk = "dingtalk",
    WeCom = "wecom"
}

/**
 * 告警严重等级
 */
export enum AlertSeverity {
    Low = "low",
    Medium = "medium",
    High = "high",
    Critical = "critical"
}

/**
 * 条件运算符
 */
export enum ConditionOperator {
    GreaterThan = ">",
    GreaterThanOrEqual = ">=",
    LessThan = "<",
    LessThanOrEqual = "<=",
    Equal = "==",
    NotEqual = "!=",
}

/**
 * 告警通道配置基础
 */
export interface BaseAlertChannelConfig {
    // 所有通道共有的字段
    [key: string]: string | string[] | boolean | Record<string, string> | undefined
}

/**
 * Webhook 通道配置
 */
export interface WebhookChannelConfig extends BaseAlertChannelConfig {
    url: string
    method?: string
    headers?: Record<string, string>
}

/**
 * Slack 通道配置
 */
export interface SlackChannelConfig extends BaseAlertChannelConfig {
    token: string
    channel: string
}

/**
 * Discord 通道配置
 */
export interface DiscordChannelConfig extends BaseAlertChannelConfig {
    webhookUrl: string
    username?: string
    avatarUrl?: string
}

/**
 * 钉钉通道配置
 */
export interface DingTalkChannelConfig extends BaseAlertChannelConfig {
    accessToken: string
    secret?: string
    atMobiles?: string[]
    isAtAll?: boolean
}

/**
 * 企业微信通道配置
 */
export interface WeComChannelConfig extends BaseAlertChannelConfig {
    webhookKey: string
    mentionedList?: string[]
    mentionedMobileList?: string[]
}

/**
 * 告警通道配置（统一类型）
 */
export type AlertChannelConfig = 
    | WebhookChannelConfig 
    | SlackChannelConfig 
    | DiscordChannelConfig 
    | DingTalkChannelConfig 
    | WeComChannelConfig
    | Record<string, string | string[] | boolean | Record<string, string> | undefined>

/**
 * 告警通道
 */
export interface AlertChannel {
    id: string
    name: string
    type: AlertChannelType
    config: AlertChannelConfig
    enabled: boolean
    createdAt: string
    updatedAt: string
}

/**
 * 告警条件
 */
export interface AlertCondition {
    metric: string
    operator: ConditionOperator
    threshold: number
    duration: number // 持续时间(分钟)
}

/**
 * 告警规则
 */
export interface AlertRule {
    id: string
    name: string
    description?: string
    conditions: AlertCondition[]
    logic: "AND" | "OR"
    channels: string[]
    template: string
    cooldown: number // 冷却时间（分钟）
    severity: AlertSeverity
    enabled: boolean
    createdAt: string
    updatedAt: string
    createdBy?: string
}

/**
 * 告警历史状态
 */
export enum AlertHistoryStatus {
    Pending = "pending",
    Sent = "sent",
    Failed = "failed",
    Acknowledged = "acknowledged"
}

/**
 * 告警历史
 */
export interface AlertHistory {
    id: string
    ruleId: string
    ruleName: string
    severity: AlertSeverity
    message: string
    details: Record<string, unknown>
    channels: string[]
    status: AlertHistoryStatus
    errorMessage?: string
    acknowledgedBy?: string
    acknowledgedAt?: string
    triggeredAt: string
    sentAt?: string
}

/**
 * 告警历史分页响应
 */
export interface AlertHistoryListResponse {
    items: AlertHistory[]
    total: number
    page: number
    pageSize: number
}

/**
 * 创建告警通道请求
 */
export interface CreateAlertChannelRequest {
    name: string
    type: AlertChannelType
    config: AlertChannelConfig
    enabled: boolean
}

/**
 * 更新告警通道请求
 */
export interface UpdateAlertChannelRequest {
    name?: string
    config?: AlertChannelConfig
    enabled?: boolean
}

/**
 * 创建告警规则请求
 */
export interface CreateAlertRuleRequest {
    name: string
    description?: string
    conditions: AlertCondition[]
    logic: "AND" | "OR"
    channels: string[]
    template: string
    cooldown: number
    severity: AlertSeverity
    enabled: boolean
}

/**
 * 更新告警规则请求
 */
export interface UpdateAlertRuleRequest {
    name?: string
    description?: string
    conditions?: AlertCondition[]
    logic?: "AND" | "OR"
    channels?: string[]
    template?: string
    cooldown?: number
    severity?: AlertSeverity
    enabled?: boolean
}

/**
 * 告警通道列表响应
 */
export interface AlertChannelListResponse {
    items: AlertChannel[]
    total: number
}

/**
 * 告警规则列表响应
 */
export interface AlertRuleListResponse {
    items: AlertRule[]
    total: number
}

/**
 * 测试告警通道请求
 */
export interface TestAlertChannelRequest {
    message: string
}

/**
 * 确认告警请求
 */
export interface AcknowledgeAlertRequest {
    comment?: string
}

/**
 * 告警统计
 */
export interface AlertStatisticsResponse {
    totalAlerts: number
    alertsBySeverity: Record<string, number>
    alertsByStatus: Record<string, number>
    // 这里只定义前端当前用到的字段；TopAlertRules 和 RecentAlerts 如需使用可再扩展
}
